src/Stiefel.jl

```diff
@@ -0,0 +1,199 @@
+    end
+    c = x'*x
+    if !isapprox(c, one(c); kwargs...)
+        return DomainError(norm(c-one(c)),
+            "The point $(x) does not lie on the Stiefel manifold of dimension (m,n), because x'x is not the unit matrix.")
+    end
+end
+
+function check_tangent_vector(S::Stiefel{M,N,T},x,v; kwargs...) where {M,N,T}
+    t = check_manifold_point(S,x)
+    if (t != nothing)
+        return t
+    end
+    if any( size(v) != representation_size(S) )
+        return DomainError(size(v),
+            "The matrix $(v) is does not lie in the tangent space of $(x) on the Stiefel manifold of dimension (m,n), since its dimensions are wrong.")
+    end
+    if !isapprox(x'*v + v'*x, zeros(N,N); kwargs...)
+        return DomainError(norm(x'*v + v'*x),
+            "The matrix $(v) is does not lie in the tangent space of $(x) on the Stiefel manifold of dimension (m,n), since x'v + v'x is not the zero matrix.")
+    end
+end
+@doc doc"""
+    exp!(M,y, x, v)
+
+compute the exponential map on the [`Stiefel`](@ref)`{m,n,T}`() manifold `M`
+eminating from `x` into tangent direction `v` and store the result in `y`.
+The formula reads
+
+$\operatorname{exp}_{x} v = \begin{pmatrix}
+   x\\v
+ \end{pmatrix}
+ \exp
+ \left(
+ \begin{pmatrix} {\bar x}^{\mathrm{T}}v & -{\bar \xi}^{\mathrm{T}}v\\
+ I_n & {\bar x}^{\mathrm{T}}v\end{pmatrix}
+ \right)
+\begin{pmatrix}  \exp( -{\bar x}^{\mathrm{T}}v) \\ 0_n\end{pmatrix}$
+
+where $\exp$ denotes matrix exponential, and $I_n$ and
+$0_n$ are the identity matrix and the zero matrix of dimension $n \times n$,
+respectively.
+"""
+function exp!(M::Stiefel{m,n,T}, y, x, v) where {m,n,T}
+    y .= [x v] * exp([x'v-v'*v; one(zeros(T,n,n)) x'*v]) * [exp(-x'v); zeros(T,n,n)]
+    return y
+end
+
+@doc doc"""
+    inverse_retract!(M, v, x, y, ::PolarInverseRetraction)
+
+compute the inverse retraction based on a singular value decomposition
+for two points `x`, `y` on the [`Stiefel`](@ref) manifold `M` and return
+the resulting tangent vector in `v`.
+The formula reads
+````math
+v = ys-x$,
+````
+where $s$ is the solution to the Lyapunov equation
+````math
+$x^{\mathrm{T}}ys + s(x^{\mathrm{T}}y)^{\mathrm{T}} + 2\mathrm{I}_k = 0.
+````
+"""
+function inverse_retract!(::Stiefel{M,N,T}, v, x, y, ::PolarInverseRetraction) where {M,N,T}
+  v .= y*lyap(x'*y, -2*one(x'*x)) - x
+  return v
+end
+
+@doc doc"""
+    inverse_retract!(M, v, x, y, ::QRInverseRetraction)
+
+compute the inverse retraction based on a qr decomposition
+for two points `x`, `y` on the [`Stiefel`](@ref) manifold `M` and return
+the resulting tangent vector in `v`. The computation follows Algorithm 1 in
+
+> T. Kaneko, S. Fiori, T. Tanaka: "Empirical Arithmetic Averaging over the
+> Compact Stiefel Manifold", IEEE Transactions on Signal Processing, 2013,
+> doi: [10.1109/TSP.2012.2226167](https://doi.org/10.1109/TSP.2012.2226167).
+"""
+function inverse_retract!(::Stiefel{M,N,T}, v, x, y, ::QRInverseRetraction) where {M,N,T}
+  A = x'*y
+  R = zeros(T,N,N)
```

Comment:
I'd replace `T` with, probably, something like `typeof(one(eltype(x))*one(eltype(y))` to properly mix types of arguments.